diff --git a/src/Plugin/views/field/SearchApiStandard.php b/src/Plugin/views/field/SearchApiStandard.php
index cb1356d5..9888d794 100644
--- a/src/Plugin/views/field/SearchApiStandard.php
+++ b/src/Plugin/views/field/SearchApiStandard.php
@@ -4,6 +4,7 @@ namespace Drupal\search_api\Plugin\views\field;
 
 use Drupal\views\Plugin\views\field\FieldPluginBase;
 use Drupal\views\Plugin\views\field\MultiItemsFieldHandlerInterface;
+use Drupal\views\Render\ViewsRenderPipelineMarkup;
 
 /**
  * Provides a default handler for fields in Search API Views.
@@ -21,7 +22,12 @@ class SearchApiStandard extends FieldPluginBase implements MultiItemsFieldHandle
    */
   public function render_item($count, $item) {
     $type = !empty($this->definition['filter_type']) ? $this->definition['filter_type'] : 'plain';
-    return $this->sanitizeValue($item['value'], $type);
+    if ($this->realField == 'rendered_item') {
+      return ViewsRenderPipelineMarkup::create($item['value']);
+    }
+    else {
+      return $this->sanitizeValue($item['value'], $type);
+    }
   }
 
 }
